File: to_delete/matmul_benchmark.py
import os
import numpy as np
import time

# Force single-threaded execution
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["OPENBLAS_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"

# Optional: Set OPENBLAS_CORETYPE before running to simulate older hardware:
#   OPENBLAS_CORETYPE=HASWELL     -> AVX2 only (~200ms)
#   OPENBLAS_CORETYPE=SANDYBRIDGE -> AVX only (~250ms)
#   OPENBLAS_CORETYPE=NEHALEM     -> SSE4.2 (~500ms)
#   OPENBLAS_CORETYPE=PRESCOTT    -> SSE3 (~530ms)
# Example: OPENBLAS_CORETYPE=PRESCOTT python matmul_benchmark.py

try:
    import threadpoolctl
    threadpoolctl.threadpool_limits(1)
except ImportError:
    pass

coretype = os.environ.get("OPENBLAS_CORETYPE", "auto-detect (likely AVX-512)")
print(f"NumPy version: {np.__version__}")
print(f"OPENBLAS_CORETYPE: {coretype}")
print(f"Threads: 1")
print()

# Create two random 2000x2000 matrices
print("Creating random 2000x2000 matrices...")
A = np.random.rand(2000, 2000)
B = np.random.rand(2000, 2000)

# Warmup round
print("Warmup round...")
_ = A @ B

# Timed rounds
times = []
for i in range(3):
    start = time.perf_counter()
    _ = A @ B
    end = time.perf_counter()
    elapsed = end - start
    times.append(elapsed)
    print(f"Round {i + 1}: {elapsed:.4f} seconds")

avg = np.mean(times)
gflops = 2 * 2000**3 / avg / 1e9
print(f"\nAverage: {avg:.4f} seconds")
print(f"Performance: {gflops:.1f} GFLOPS")

File: to_delete/matmul_benchmark.mjs
/**
 * Matrix multiplication benchmark for numwasm
 * Tests matmul on two random 2000x2000 matrices
 */

import * as nw from "../packages/numwasm/dist/numwasm.mjs";

const N = 2000;
const ROUNDS = 3;

async function benchmark() {
  // Load WASM module first
  console.log("Loading WASM module...");
  await nw.loadWasmModule();
  console.log("WASM module loaded.\n");

  console.log(`Creating two random ${N}x${N} matrices...`);

  // Create random matrices using rand
  const createStart = performance.now();
  const a = nw.rand(N, N);
  const b = nw.rand(N, N);
  const createEnd = performance.now();
  console.log(
    `Matrix creation time: ${(createEnd - createStart).toFixed(2)} ms`,
  );

  console.log(`Matrix A shape: ${a.shape}`);
  console.log(`Matrix B shape: ${b.shape}`);

  // Warmup round
  console.log(`\n--- Warmup round ---`);
  const warmupStart = performance.now();
  await nw.linalg.matmul(a, b);
  const warmupEnd = performance.now();
  console.log(`Warmup time: ${(warmupEnd - warmupStart).toFixed(2)} ms`);

  // Timed rounds
  const times = [];
  for (let i = 0; i < ROUNDS; i++) {
    console.log(`\n--- Round ${i + 1} ---`);
    const start = performance.now();
    const result = await nw.linalg.matmul(a, b);
    const end = performance.now();
    const elapsed = end - start;
    times.push(elapsed);
    console.log(`Time: ${elapsed.toFixed(2)} ms`);
  }

  // Calculate average
  const avgTime = times.reduce((a, b) => a + b, 0) / times.length;
  const flops = 2 * N * N * N;
  const gflops = flops / (avgTime / 1000) / 1e9;

  console.log(`\n========== Results ==========`);
  console.log(`Matrix size: ${N}x${N}`);
  console.log(`Rounds: ${ROUNDS}`);
  console.log(`Times: ${times.map((t) => t.toFixed(2) + " ms").join(", ")}`);
  console.log(`Average time: ${avgTime.toFixed(2)} ms`);
  console.log(`Performance: ${gflops.toFixed(2)} GFLOPS`);
}

benchmark().catch(console.error);
